src/visitors/semantic/symbols/mod.rs

```diff
@@ -0,0 +1,198 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::protocol::properties::{SymbolInformation, SymbolKind};
+
+use flux::semantic::nodes::Expression;
+use flux::semantic::walk::{Node, Visitor};
+
+#[derive(Default)]
+pub struct SymbolsState<'a> {
+    pub symbols: Vec<SymbolInformation>,
+    pub uri: String,
+    pub path: Vec<Rc<Node<'a>>>,
+}
+
+#[derive(Default)]
+pub struct SymbolsVisitor<'a> {
+    pub state: Rc<RefCell<SymbolsState<'a>>>,
+}
+
+impl<'a> SymbolsVisitor<'a> {
+    pub fn new(uri: String) -> SymbolsVisitor<'a> {
+        let state = SymbolsState {
+            path: vec![],
+            symbols: vec![],
+            uri,
+        };
+        SymbolsVisitor {
+            state: Rc::new(RefCell::new(state)),
+        }
+    }
+}
+
+impl<'a> Visitor<'a> for SymbolsVisitor<'a> {
+    fn done(&mut self, _: Rc<Node<'a>>) {
+        let mut state = self.state.borrow_mut();
+        (*state).path.pop();
+    }
+
+    fn visit(&mut self, node: Rc<Node<'a>>) -> bool {
+        let mut state = self.state.borrow_mut();
+        let uri = (*state).uri.clone();
+
+        (*state).path.push(node.clone());
+
+        match node.as_ref() {
```

Comment:
Have some cyclomatic complexity  issue. Hard to figure out the layers. maybe cut them into small funcs